### PR TITLE
[WIP] Integration with Link checker api

### PR DESCRIPTION
**Pull request by kevindew**
[Trello Card](https://trello.com/c/RkRvwP9k/969-complete-whitehall-integration)

### CODE REVIEW

File: app/controllers/admin/link_checker_api_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      report.update_from_batch_report(params) if report
+    end
+
+    head :no_content
+  end
+
+private
+
+  def verify_webhook
+    # TODO
```
Comment (thomasleese) [inline app/controllers/admin/link_checker_api_controller.rb]: Hehe

File: app/models/link_checker_api_report/link.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class LinkCheckerApiReport::Link < ActiveRecord::Base
+  serialize :check_errors, Hash
+  serialize :check_warnings, Hash
+
+  belongs_to :report, class_name: LinkCheckerApiReport
+
+  def self.attributes_from_link_report(payload)
```
Comment (thomasleese) [inline app/models/link_checker_api_report/link.rb]: Can we use the `LinkReport` model from GDS API Adapters here? Or in similar places?

File: db/schema.rb [hunk 3]
```diff
@@ -83,7 +83,6 @@
   add_index "attachments", ["attachable_id", "attachable_type"], name: "index_attachments_on_attachable_id_and_attachable_type", using: :btree
   add_index "attachments", ["attachable_type", "attachable_id", "ordering"], name: "no_duplicate_attachment_orderings", unique: true, using: :btree
   add_index "attachments", ["attachment_data_id"], name: "index_attachments_on_attachment_data_id", using: :btree
-  add_index "attachments", ["content_id"], name: "index_attachments_on_content_id", using: :btree
```
Comment (thomasleese) [inline db/schema.rb]: Wonder why this happened here 🤔

File: test/functional/admin/link_check_reports_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    login_as :writer
+    @publication = create(:publication, body: "[link](http://www.example.com)")
+    Rails.backtrace_cleaner.remove_silencers!
+
+    link_checker_endpoint = "#{Plek.find('link-checker-api')}/batch"
+    body = link_checker_api_batch_report_hash(
+      id: 5,
+      links: [{ uri: "http://www.example.com" }],
+    )
+    stub_request(:post, %r{\A#{link_checker_endpoint}})
```
Comment (thomasleese) [inline test/functional/admin/link_check_reports_controller_test.rb]: There are some helper methods which can stub the whole request, have you tried those?

File: db/schema.rb (hunk 3 above)
Comment (kevindew) [inline db/schema.rb]: oh yes that's not good - good spot

File: app/models/link_checker_api_report/link.rb (hunk 2 above)
Comment (kevindew) [inline app/models/link_checker_api_report/link.rb]: This uses the output of the webhook as well so it's designed to be compatible with both a strong params object and the output from GDS API - not sure if that might bite at some point though.

File: test/functional/admin/link_check_reports_controller_test.rb (hunk 4 above)
Comment (kevindew) [inline test/functional/admin/link_check_reports_controller_test.rb]: Whitehall uses a different hostname (alphagov.co.uk) than is stubbed by GDS API adapters (dev.gov.uk) which is super annoying.

File: app/models/link_checker_api_report/link.rb (hunk 2 above)
Comment (thomasleese) [inline app/models/link_checker_api_report/link.rb]: I was thinking actually turning the input from the web hook into a `BatchReport` object and then just using that -- it's what we use in LLM.

**Comment by kevindew** [comment]
I'll close this for now since it's not being actively worked on.

**Comment by thomasleese** [comment]
Will be working on this during the fire break.

**Comment by rubenarakelyan** [comment]
Closing for the moment since it's not being actively worked on.

**Pull request closed without merging**